### PR TITLE
vSphere infrastructure provider template

### PR DESCRIPTION
**Pull request by eaudetcobello**
Gaps:

- `insecure-flag: true` issue and sha1/sha256 thumbprint needs to be investigated. TLDR, there's conflicting documentation where the provider says the thumbprint needs to be in sha256 format, but providing the thumbprint in sha256 causes a "thumbprint mismatch" error. I have resolved this by using a sha1 thumbprint *and* setting `insecure-flag: true`. Need to try *only* setting `insecure-flag: true` with sha256, or removing `insecure-flag: true` and use sha1.

**NOTE** Do not merge before testing with multi-node cluster.
**NOTE** Verify if L297-301, L319-323, L293-295 are needed for multi-node cluster before merging.

### CODE REVIEW

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:12]: let's use the same CIDRs we have in other places, `10.1.0.0/16` for pods and `10.152.183.0/24` for services

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:185]: you should create the RBAC rules and use that serviceaccount instead of mounting the admin kubeconfig

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:220]: we don't need this, the cloud-init files will create intermediate directories as needed

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:221]: sudo is not needed, add a NOTE why this is here (this needs to go away before merging)

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:234]: from here and other resources, remove the `namespace:` (NOTE: do not remove it from the cloud-provider manifests). the namespace is decided by the `clusterctl generate cluster --target-namespace $ns` argument

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:256]: this should be ${WORKER_MACHINE_COUNT}

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:108]: two notes here:

- best move these to a secret (e.g. `${CLUSTER_NAME}-resources`) and then use `contentFrom` here to keep this section clean and manageable
- for the proxy configuration specifically, we cannot hardcode it into the template (having a vsphere environment is orthogonal to requiring a proxy). maybe you could add a ${VSPHERE_PROXY_DISABLED} prefix on all lines here, then add it to the list of vars as empty (to enable) or `# ` to disable.
-

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:227]: this should be ${CONTROL_PLANE_MACHINE_COUNT}

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [inline templates/vsphere/template-variables.rc:20]: let's set the whole image as an argument here, same for csi

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [inline templates/vsphere/template-variables.rc:2]: add CONTROL_PLANE_MACHINE_COUNT and WORKER_MACHINE_COUNT vars

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [inline templates/vsphere/template-variables.rc:4]: let's avoid this

**Review by neoaggelos — CHANGES_REQUESTED** [review]
Immaculate work getting it to work! Let's clean up the rough edges!

File: templates/vsphere/template-variables.rc
Comment (eaudetcobello) [reply templates/vsphere/template-variables.rc:4]: Avoid what?

Comment (neoaggelos) [reply templates/vsphere/template-variables.rc:4]: Let's avoid fake credentials. This should be empty or commented out, so that if people forget to add their password, they will fail to generate a cluster template altogether (rather than getting a cluster template with a placeholder password)

File: templates/vsphere/template-variables.rc
Comment (eaudetcobello) [reply templates/vsphere/template-variables.rc]: Smart

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [inline templates/vsphere/template-variables.rc:31]: the standard name for that is `KUBERNETES_VERSION`, which comes from `clusterctl generate cluster --kubernetes-version`

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [inline templates/vsphere/template-variables.rc:15]: same here, leave these unset, and add a `, e.g. "10.0.0.1"` etc at the end of the comments

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [inline templates/vsphere/template-variables.rc:36]: newline at EOF

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [inline templates/vsphere/template-variables.rc:33]: this is not `VSPHERE_` related.

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [inline templates/vsphere/template-variables.rc:17]: remember, this could be optional

**Review by neoaggelos — COMMENTED** [review]
Looks great, did another pass

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [reply templates/vsphere/template-variables.rc:15]: this is to prevent issues where this config is not set properly, and instead of a failure users get an invalid cluster template

**Review by bschimke95 — COMMENTED** [review]
LGTM, waiting for @neoaggelos final approval

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [inline templates/vsphere/template-variables.rc:31]: move the kubernetes version at the top, followed by control plane and worker node counts. then, all our cluster templates can be similar.

the order should be version -> cluster size -> vsphere configs -> ck8s configs

File: templates/vsphere/template-variables.rc
Comment (neoaggelos) [inline templates/vsphere/template-variables.rc:33]: we cannot hardcode the proxy to squid.internal. Perhas move all the configs in here, as an example

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:301]: double check that these are required. for the authorized keys, maybe a `files` entry would be the better option.

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:271]: same here

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:272]: we should not be promoting this for canonical k8s. instead, either override the `install.sh` script, or wait for a stable version before merging the PR.

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:13]: we should not hardcode this. Consider whether we could have an option like `PROXY_CONFIGURATION` with a value like

```
PROXY_CONFIGURATION="
http_proxy=...
https_proxy=...
no_proxy=....
"
```

Then use `[Service] EnvironmentFile` and point to that file instead.

Comment (neoaggelos) [reply templates/vsphere/cluster-template.yaml:13]: See https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Command%20lines for the `EnvironmentFile` directive

**Review by neoaggelos — CHANGES_REQUESTED** [review]

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:255]: move these files to a secret, then specify `contentFrom`, such that CK8sControlPlane resource remains tidy

File: templates/vsphere/cluster-template.yaml
Comment (neoaggelos) [inline templates/vsphere/cluster-template.yaml:259]: validate that this is required

File: templates/vsphere/cluster-template.yaml
Comment (eaudetcobello) [reply templates/vsphere/cluster-template.yaml]: Keys moved to files. Will add a note to verify the preRunCommands before merging.

File: templates/vsphere/cluster-template.yaml
Comment (eaudetcobello) [reply templates/vsphere/cluster-template.yaml:296]: Will verify when we can create multi-node clusters.

File: templates/vsphere/template-variables.rc
Comment (bschimke95) [inline templates/vsphere/template-variables.rc:32]: for AWS you don't set defaults. Please either update the AWS env (in another PR) or this rc file
https://github.com/canonical/cluster-api-k8s/blob/main/templates/aws/template-variables.rc

Comment (eaudetcobello) [reply templates/vsphere/template-variables.rc:32]: @bschimke95 Not sure what you mean here

Comment (bschimke95) [reply templates/vsphere/template-variables.rc:32]: You don't set an empty string in the AWS template e.g. `export AWS_PUBLIC_IP=`

**Review by bschimke95 — APPROVED** [review]
LGTM

**Pull request merged**
